c_emulator/riscv_softfloat.c

```diff
@@ -15,9 +16,70 @@ static uint_fast8_t uint8_of_rm(mach_bits rm)
   softfloat_roundingMode = (uint_fast8_t)rm
 
 #define SOFTFLOAT_POSTLUDE(res)                                                \
-  zfloat_result = res.v;                                                       \
+  recreate_lbits_of_fbits(&zfloat_result, res.v, 128ull, true);                \
   zfloat_fflags = (mach_bits)softfloat_exceptionFlags
 
+static float128_t to_float128(sail_bits v1)
+{
+  static mpz_t tmp;
+  static bool isInitialized = false;
+
+  float128_t res;
+  uint32_t lo;
+  uint32_t hi;
+  mpz_t *from = v1.bits;
+
+  // I don't know if you mind a small memory leak or not
+  // malloc not being released
+  if (!isInitialized) {
+    mpz_init(tmp);
+    isInitialized = true;
+  }
+
+  mpz_mod_2exp(tmp, *from, 64); // lower 64 bits of v1
```

Comment:
I'm not super familiar with the more obscure Sail functions, but I know Sail implements a bunch of the mpz functions. Is there a way to do this directly with Sail datatypes and functions instead of having to create mpz_t variables? Would love any guidance here. Maybe @Alasdair

Reply:
> Yeah I think it would be much much much simpler just to split it into two 64 bit values in Sail and pass those in as two separate parameters. They will arrive in C as `uint64_t`.

Correct me if I am wrong. I think by the time when I was implementing c to sail interface, one issue is that sail memory model only support at most 64bit memory transaction, therefore, I splitted quad word floating point load and store as two different memory movement.

I am not sure if this would cause any concurrency problem.

Reply:
I've changed it so that the functions in riscv_softfloat_interface.sail pass two 64 bit values to the external softfloat implementation. The riscv_softfloat.c file puts them together for the softfloat calls themselves and then splits them apart again to return them to Sail as 64 bit values. Let me know what you think. This eliminates the need for mpz entirely.